### ISSUE TITLE
Softmax Approximation for Large Output Layers (size of experts)

### ISSUE THREAD

**Issue by hosseinfani** [fani-lab/OpeNTF#168]
Our current neural architecture uses softmax at the output layers whose size is the size of the expert set. 
We need to provide a faster approximation to softmax while not loosing performance.

hierarchical softmax,  svd-softmax, ... are available options.

**Comment by hosseinfani** [comment]
@rezaBarzgar

https://github.com/fani-lab/OpeNTF/blob/148c1c2defe1176563f162ad159b2ffe0af15ecc/src/mdl/fnn.py#L36

**Comment by rezaBarzgar** [comment]
### The answer to the question that you asked in the meeting (9 Oct 2022):

Difference between sigmoid & softmax:
outputs of the sigmoid function are independent while those of softmax are interrelated. In other words, the sum of softmax outputs is always 1, but it is incorrect for sigmoid. Increase in probability of one class results in decrease of other class or classes probability.

### questions

- I ran OpeNTF with dblp toy on fnn with softmax activation function instead of sigmoid. Could you explain why sigmoid is used in fnn model in OpeNTF?
- Where can I find papers about hierarchical softmax?

@hosseinfani

**Comment by hosseinfani** [comment]
@rezaBarzgar 
Thanks for the update and sorry for the late reply. 

I reviewed the code and found that we select sigmoid because we have a multiclass classification, each member being a class, that is, an input skill can belong to multiple members to form the team.

So, this task can be closed. No need for speedup for softmax for this architecture. We can start another task.

Before that, can you please study whether the sigmoid has been applied on all nodes of the output layer in parallel?

**Comment by rezaBarzgar** [comment]
@hosseinfani 
Sorry, the question is a little bit ambiguous for me.
If your question is "Is sigmoid applied on all nodes in the output layer?", I believe the answer is yes.
But, if your question is "if we run OpeNTF in parallel mode, has sigmoid been applied on all nodes of the output layer?", I could not find where to look for it.

**Comment by hosseinfani** [comment]
@rezaBarzgar 
my question is "Is sigmoid applied on all nodes in the output layer *in parallel*?" Not sure my question is sound tho.

like in below sample, the sigmoid is applied on `input` elements in parallel?

```
import torch
input = torch.randn(20)
torch.sigmoid(input)
```

**Comment by hosseinfani** [comment]
@rezaBarzgar 
Just a reminder that we used pytorch, not tensorflow. But I think they follow the same idea in underlying functions. 
We can safely close this issue then.
